src/main/java/org/opensearch/neuralsearch/search/collector/HybridTopFieldDocSortCollector.java

```diff
@@ -0,0 +1,403 @@
+        this.numHits = numHits;
+        this.hitsThresholdChecker = hitsThresholdChecker;
+        this.sort = sort;
+    }
+
+    /**
+     * HybridCollectorManager fetches the topDocs in the reduce method.
+     * @return List of TopFieldDocs which represents results of Top Docs of individual subquery.
+     */
+    public List<TopFieldDocs> topDocs() {
+        if (compoundScores == null) {
+            return new ArrayList<>();
+        }
+
+        List<TopFieldDocs> topFieldDocs = new ArrayList<>();
+        for (int subQueryNumber = 0; subQueryNumber < compoundScores.length; subQueryNumber++) {
+            topFieldDocs.add(
+                topDocsPerQuery(
+                    0,
+                    Math.min(collectedHits[subQueryNumber], compoundScores[subQueryNumber].size()),
+                    compoundScores[subQueryNumber],
+                    collectedHits[subQueryNumber],
+                    sort.getSort()
+                )
+            );
+        }
+        return topFieldDocs;
+    }
+
+    @Override
+    public ScoreMode scoreMode() {
+        return hitsThresholdChecker.scoreMode();
+    }
+
+    protected abstract class HybridTopDocSortLeafCollector implements LeafCollector {
+        protected HybridQueryScorer compoundQueryScorer;
+        private boolean collectedAllCompetitiveHits = false;
+        @Nullable
+        private FieldDoc after;
+        private final Sort sort;
+        private boolean initializePerSegment;
+
+        public HybridTopDocSortLeafCollector(Sort sort, @Nullable FieldDoc after) {
+            this.sort = sort;
+            this.after = after;
+            this.initializePerSegment = true;
+        }
+
+        @Override
+        public void setScorer(final Scorable scorer) throws IOException {
+            if (scorer instanceof HybridQueryScorer) {
+                log.debug("passed scorer is of type HybridQueryScorer, saving it for collecting documents and scores");
+                compoundQueryScorer = (HybridQueryScorer) scorer;
+            } else {
+                compoundQueryScorer = getHybridQueryScorer(scorer);
+                if (Objects.isNull(compoundQueryScorer)) {
+                    log.error(
+                        String.format(Locale.ROOT, "cannot find scorer of type HybridQueryScorer in a hierarchy of scorer %s", scorer)
+                    );
+                }
+            }
+        }
+
+        private HybridQueryScorer getHybridQueryScorer(final Scorable scorer) throws IOException {
+            if (scorer == null) {
+                return null;
+            }
+            if (scorer instanceof HybridQueryScorer) {
+                return (HybridQueryScorer) scorer;
+            }
+            for (Scorable.ChildScorable childScorable : scorer.getChildren()) {
+                HybridQueryScorer hybridQueryScorer = getHybridQueryScorer(childScorable.child);
```

Comment:
so is this idea of getting one child score , is that reason of recursive call?

Reply:
Yes because it can be nested.